File: cpp/131分割回文串/131分割回文串.cpp
/*
给定一个字符串 s，将 s 分割成一些子串，使每个子串都是回文串。

返回 s 所有可能的分割方案。

示例:

输入: "aab"
输出:
[
  ["aa","b"],
  ["a","a","b"]
]

*/

//解答：递归，每次分割成功一个子串后，对剩下来的子串递归分割，一直到分割完则加入结果中
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        if(s.empty())
            return result;
        vector<string> path;
        solve(result, path, s, 0);
        return result;
    }
    
    void solve(vector<vector<string>>& result, vector<string>& path, string& s, int pos)
    {
        if(pos == s.length())
        {
            result.push_back(path);
            return;
        }
        
        for(int i = pos; i < s.length(); i++)
        {
            string temp = s.substr(pos, i - pos + 1);
            if(!isPalindrome(temp))
                continue;
            
            path.push_back(temp);
            solve(result, path, s, i + 1);
            path.pop_back();
        }
    }
    
    bool isPalindrome(string& s)
    {
        if(s.length() == 0)
            return false;
        int i = 0, j = s.length() - 1;
        while(i <= j)
        {
            if(s[i++] != s[j--])
                return false;
        }
        
        return true;
    }
};

File: python/131分割回文串/131分割回文串.py
"""
给定一个字符串 s，将 s 分割成一些子串，使每个子串都是回文串。

返回 s 所有可能的分割方案。

示例:

输入: "aab"
输出:
[
  ["aa","b"],
  ["a","a","b"]
]

"""

# 解答：递归，每次分割成功一个子串后，对剩下来的子串递归分割，一直到分割完则加入结果中
class Solution:
    def partition(self, s: str):
        if len(s) == 0:
            return []
        self.result = []
        self.path = []
        self.solve(s, 0)
        return self.result
    
    def solve(self, s, pos):
        if pos == len(s):
            self.result.append(self.path[:])
        
        for i in range(pos, len(s)):
            temp = s[pos:i+1]
            if not self.isPalindrome(temp):
                continue
            
            self.path.append(temp)
            self.solve(s, i + 1)
            self.path.pop()
        
    
    def isPalindrome(self, s):
        if len(s) == 0:
            return False
        
        begin = 0
        end = len(s) - 1
        while begin < end:
            if s[begin] != s[end]:
                return False
            begin += 1
            end -= 1
        
        return True